### PR TITLE
Store pubkey bin offsets in hash cache file

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem

Storing pubkey bin offsets in hash cache file to avoid recomputing them when deduping. 

65K * 8 * 300K = 156G data

master 
```
sol@dev-equinix-washington-23:~$ grep calculate_accounts_hash_from_storages d0 | tail -n1 | awk '{for(i=1; i<=NF; i++) if (0 != index($i, "=")) {print($i)}}' | column -ts "=" | grep -E "pubkey_bin_search_us|eliminate_zeros_us|min_pk_scan_us|hash_writes_us|skip_duplicates_us|^sort_us|hash_total|total_us"
total_us                            9190042i
eliminate_zeros_us                  4677618i
sort_us                             674877i
hash_total                          338924320i
pubkey_bin_search_us                12422676i
```

this PR
```
sol@dev-equinix-washington-23:~$ grep calculate_accounts_hash_from_storages d2 | tail -n1 | awk '{for(i=1; i<=NF; i++) if (0 != index($i, "=")) {print($i)}}' | column -ts "=" | grep -E "pubkey_bin_search_us|eliminate_zeros_us|min_pk_scan_us|hash_writes_us|skip_duplicates_us|^sort_us|hash_total|total_us"
total_us                            7825945i
eliminate_zeros_us                  3338763i
sort_us                             678038i
hash_total                          338924320i
pubkey_bin_search_us                0i
```

Cut down de-dup by 1.3s. 

Dedup timing on mainnet

image

Red: master; blue: this PR.

#### Summary of Changes

Fixes #

### CODE REVIEW

**Comment by jeffwashington** [comment]
`65K * 8 * 300K = 156G data`

Where does `300K` come from? We should not have this many chunks unless the chunking algorithm is broken.

**Comment by jeffwashington** [comment]
> `65K * 8 * 300K = 156G data`

We could do fancy things to attack the `8` such as using u32s. There can never be 4B entries in a file. That would require more than 16G of file size, which is the max append vec size at least. That certainly helps. At a minimum, the file contains a pubkey and hash, which are both not compressible.

**Comment by jeffwashington** [comment]
`calculate_accounts_hash_from_storages.scan_chunks` gives you the # of chunks (the 300k number above I think).

**Comment by HaoranYi** [comment]
> Where does `300K` come from?

That's the worst case, we have seen from pop.

**Comment by HaoranYi** [comment]
yeah. u32 sounds good to me.

**Comment by jeffwashington** [comment]
> > Where does `300K` come from?
> 
> That's the worst case, we have seen from pop.

Here is the most recent pop 512 run:
image

The worst case from above was at 20B accounts, where ancient packing was almost certainly not keeping up.
we are seeing 40k, 70k, 90k (70 and 90 are rpc machines). Not sure what's going on there. Also not sure why the other 3 staked nodes aren't reporting at all...

**Comment by jeffwashington** [comment]
> > `65K * 8 * 300K = 156G data`
> 
> We could do fancy things to attack the `8` such as using u32s. There can never be 4B entries in a file. That would require more than 16G of file size, which is the max append vec size at least. That certainly helps. At a minimum, the file contains a pubkey and hash per account, which are both not compressible.

 I don't know if we can get down to 1 less byte (u24). I estimate 1.2B bytes for u24::MAX (16,777,216) * 72 bytes. That is less than the current ancient packing target of ~1.6GB. So, we'd have to special case if we got a range that had > 16M pubkeys from a single bin (such as if we had only 1 bin). Doable, just more complicated, but provides a savings of 1 byte * # bins (65k) per file (90k max in pathological case atm). This would save 5.8GB on disk. But, again, if we are at 90k chunks, we need to change the chunking algorithm or change ancient packing to work better.

**Comment by jeffwashington** [comment]
Note that we have to invalidate the old files so we don't crash somehow. Maybe in the file name or in the hash we calculate to create the file names.

File: accounts-db/src/accounts_db.rs [hunk 1]
```diff
@@ -7250,7 +7250,7 @@ impl AccountsDb {
                     // so, build a file name:
                     let hash = hasher.finish();
                     let file_name = format!(
-                        "{}.{}.{}.{}.{:016x}",
+                        "new.{}.{}.{}.{}.{:016x}",
```
Comment (jeffwashington) [inline accounts-db/src/accounts_db.rs:7253]: I'd prefer `.1` added to end. Make `.1` a version constant
`const VERSION: u64 = 1;`
and pass VERSION for new {} arg at end.
`new` ages itself ;-)

Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: yeah. done.

**Pull request closed without merging**